### DISCUSSION TITLE
PIco in host mode works but I need to declare a global variable

### DISCUSSION THREAD

**Discussion by danjperron** [micropython/micropython#9235] [Core Development] [answered]
OK I was able to switch micropython in usb host mode to get a USB gamepad working.

I want to get ridd of the Pi to read the gamepad and send udp datagram to my small robot!

https://www.youtube.com/watch?v=nC8gTRUxHxI

I just add some c funtion to poll the usbhost to the the gamepad.  It was working !!! yeah

But I  made a mistake  and now I can't remember how I declare a  a 10 byte variable.

I was using MP_REGISTER_ROOT_POINTER(uint8_t gamepad[10]) and MP_STATE_VM(gamepad) to set and get the variable but now micropython can't compile anymore. It was working before. Did I use the correct stuff to implement my variable

Question 1 .  

   How to I declare a global variable of 10 bytes  that my c function in gamepad module could read. I though it was MP_STATE_VM but obviously its not or I missed something.

Question 2.
 
 I create a function to poll the gamepad but I need to put in in a loop when a script is running.  Works Well if I use Thonny and not running a scrirpt.  So where is the best place to put the poll_gamepad() to empty the usb host and get my gamepad data.

compile gives error!  GamePadData is not define anymore
`static void process_gamepad_report(hid_gamepad_report_t const * report)
{
 int  sizeR = sizeof(hid_gamepad_report_t);
 //int loop;

 uint8_t * pt = (uint8_t *) MP_STATE_VM(GamePadData);

 memcpy(pt,report,10);

 //for(loop=0;loop<sizeR;loop++)
 //  printf("%02X ", (int) pt[loop]);
 // printf("\n");
 //my gamepad doesnt have  z axis so hat is on ry
 //printf("hat:0x%02X\n",report->ry);
}
`

and my gamepad.get()
`
STATIC mp_obj_t gamepad_get(void) {
//    int loop;

 uint8_t * pt = (uint8_t *) MP_STATE_VM(Game_PadData);
 //    printf("gamepad pt:%08lx  :",(uint32_t) pt);
 //   for (loop=0;loop<10;loop++)
 //     printf("%02X ", (int) pt[loop]);
 //   printf("\n");
    return mp_obj_new_bytearray_by_ref(10,pt);
`}`

```

**Comment by dlech** [comment]
> I was using MP_REGISTER_ROOT_POINTER(uint8_t gamepad[10]) and MP_STATE_VM(gamepad) to set and get the variable but now micropython can't compile anymore. It was working before. Did I use the correct stuff to implement my variable

Since `gamepad` doesn't contain any pointers to anything else, it doesn't need to be a root pointer. You should just be able to have a global `static uint8_t gamepad[10]` and use it directly. 

> compile gives error! GamePadData is not define anymore

If you are registering `gamepad` then that doesn't match `GamePadData`, so the error makes sense.

**Comment by danjperron** [comment]
oops typo on  my part.    MP_REGISTER_ROOT_POINTER(uint8_t GamePadData[10])  and  MP_STATE_VM(GamePadData) is the real term i'm using on my code.   I tried (void *) , (uint8_t *) and got the same result. 
Just try to figure out why  MP_REGISTER_ROOT_POINTER ..... was ok and suddendly  MP_STATE_VM(GamePadData) compilation telling me that GamePadData is not defined. Are both of them related?     I will just start fresh just to see if I erased something.  

is there another way to have a global variable that could be link to a module.  The example in C aren't very good to explain how to manipulate object .

**Comment by danjperron** [comment, marked as answer]
OK I found my bug!   I forget to add the file gamepad_var.c into the CMakLists.txt MICROPY_SOURCE_QSTR

this file gamepad_var.c is simple
```
//  global GamePadData
//  10  bytes  to transfer gamepad info
//  it is access from module gamepad.c and hid_app.c
//
// using,
// uint8_t * = (uint8_t *) MP_STATE_VM(GamePadData);

#include "py/mpstate.h"

MP_REGISTER_ROOT_POINTER(uint8_t GamePadData[10]);
```


Now Where do I put the poll to read the usb host in a loop to be able to get information about the gamepad when the Pico run a script.

**Reply by dpgeorge** [reply]
Look at `MICROPY_EVENT_POLL_HOOK_FAST` and possibly `MICROPY_HW_USBDEV_TASK_HOOK`

**Reply by danjperron** [reply]
Thanks Damien,

 I will look into it.  I got it working. using a timer.  I did make a external c function to get the gamepad info and then use a Python script to poll it every 100Hz.

Works fine! 
 https://github.com/danjperron/pico4legsBot/tree/main/USB_HOST

**Reply by dpgeorge** [reply]
We do want to have official support for USB host mode in MicroPython, and also more configurable USB device mode.  It's in the works!

@projectgus FYI the above USB_HOST example may be useful.

**Reply by danjperron** [reply]
Thank you very  much dpgeorge.  The MICROPY_EVENT_POLL_HOOK_FAST fix my poll problem. I don't need an extra Python script to poll the USB HID gamepad anymore.  I also removed anything on main.c

I update my github. and this is a small video of my little bot using a USB gamepad dongle.
https://youtu.be/K8TxNaSnl-0
